concordium-consensus/src/Concordium/Skov/Update.hs

```diff
@@ -162,18 +162,29 @@ processFinalization newFinBlock finRec@FinalizationRecord{..} = do
         -- We proceed backwards from the new finalized block, marking it and
         -- its ancestors as finalized, while other blocks at the same height
         -- are marked dead.
+        -- Instead of marking blocks dead immediately we accummulate them
+        -- and a return a list. The reason for doing this is that we never
+        -- have to look up a parent block that is already marked dead.
         let
-            pruneTrunk :: BlockPointerType m -> Branches m -> m ()
-            pruneTrunk _ Seq.Empty = return ()
-            pruneTrunk keeper (brs Seq.:|> l) = do
-                forM_ l $ \bp -> if bp == keeper then do
-                                    markFinalized (getHash bp) finRec
-                                    logEvent Skov LLDebug $ "Block " ++ show bp ++ " marked finalized"
-                                else do
-                                    markLiveBlockDead bp
-                                    logEvent Skov LLDebug $ "Block " ++ show bp ++ " marked dead"
+            pruneTrunk :: [BlockPointerType m] -- ^List of blocks to remove.
+                       -> BlockPointerType m -- ^The block that was finalized.
+                       -> Branches m
+                       -> m [BlockPointerType m]
+                       -- ^ The return value is a list of blocks to mark dead, ordered
+                       -- by increasing height.
+            pruneTrunk toRemove _ Seq.Empty = return toRemove
+            pruneTrunk toRemove keeper (brs Seq.:|> l) = do
+                toRemove1 <- foldM (\remove bp -> if bp == keeper then do
+                                      markFinalized (getHash bp) finRec
+                                      logEvent Skov LLDebug $ "Block " ++ show bp ++ " marked finalized"
+                                      return remove
+                                    else
+                                      return (bp:remove)
+                                  )
+                            toRemove
+                            l
                 parent <- bpParent keeper
-                pruneTrunk parent brs
+                toRemove2 <- pruneTrunk toRemove1 parent brs
```

Comment:
It might be preferable to call `markFinalized` after the recursive call to ensure that blocks are marked finalized in increasing height order. (This would also be useful for database recovery, since it would be convenient to assume that if a block is in the database then its parent will also be there.)